dlt/destinations/impl/filesystem/layout.py

```diff
@@ -0,0 +1,250 @@
+    "curr_date",
+    "year",
+    "month",
+    "day",
+    "hour",
+    "minute",
+    "dow",
+    "timestamp",
+}
+
+SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
+
+
+class ExtraParams:
+    def __init__(
+        self,
+        config: FilesystemDestinationClientConfiguration,
+        job_info: Optional[ParsedLoadJobFileName] = None,
+        schema_name: Optional[str] = None,
+        load_id: Optional[str] = None,
+    ) -> None:
+        self.config = config
+        self.job_info = job_info
+        self.load_id = load_id
+        self.schema_name = schema_name
+        self._params: Dict[str, Any] = {}
+
+        self.table_name = None
+        self.file_id = None
+        self.file_format = None
+        if job_info:
+            self.table_name = job_info.table_name
+            self.file_id = job_info.file_id
+            self.file_format = job_info.file_format
+
+    @property
+    def params(self) -> Optional[Dict[str, Any]]:
+        """Process extra params for layout
+        If any value is a callable then we call it with the following arguments
+
+            * schema name,
+            * table name,
+            * load id,
+            * file id,
+            * current datetime
+        """
+        if self.load_id:
+            self._params["load_id"] = self.load_id
+
+        if self.job_info:
+            self._params.update(
+                {
+                    "table_name": self.table_name,
+                    "file_id": self.file_id,
+                    "ext": self.file_format,
+                }
+            )
+
+        # If current_datetime is not given
+        # Then take current moment in time
+        # If current_datetime is callable
+        # Then call it and check it's instance
+        # If the result id DateTime
+        # Then take it
+        # Else exit.
+        if not self.config.current_datetime:
+            self.config.current_datetime = pendulum.now()
+        elif callable(self.config.current_datetime):
+            result = self.config.current_datetime()
+            if isinstance(result, pendulum.DateTime):
+                self.current_datetime = result
+            else:
+                raise RuntimeError(
+                    "current_datetime was passed as callable but "
+                    "didn't return any instance of pendulum.DateTime"
+                )
+
+        now: pendulum.DateTime = self.config.current_datetime  # type: ignore[assignment]
+
+        # Format curr_date datetime according to given format
+        if self.config.datetime_format:
+            self._params["curr_date"] = now.format(self.config.datetime_format)
+        else:
+            self._params["curr_date"] = str(now.date())
+
+        # For each callable extra parameter
+        # otherwise take it's value
+        if self.config.extra_params:
+            for key, value in self.config.extra_params.items():
+                if callable(value):
+                    self._params[key] = value(
```

Comment:
here you must catch TypeError which indicates a wrong signature and explain what the correct signature is. otherwise people will be confused

ideally you could inspect all the signatures when validating placeholders